### PR TITLE
New resource: azurerm_private_dns_txt_record

### LINKED ISSUE #5012: Support for azurerm_private_dns_txt_record
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Would be nice to have also resource for adding txt records to private dns.

### New or Affected Resource(s)

* azurerm_private_dns_txt_record

### Potential Terraform Configuration

```hcl

resource "azurerm_private_dns_txt_record" "example" {
  name                = "foobar"
  zone_name           = "${azurerm_private_dns_zone.example.name}"
  resource_group_name = "${azurerm_resource_group.example.name}"
  ttl                 = 300
  txt              = "foobar txt string"
}

```

### References

* #0000

### PR DESCRIPTION
**Pull request by manicminer**
Fixes #5012

### CODE REVIEW

File: azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+			"resource_group_name": azure.SchemaResourceGroupNameDiffSuppress(),
+
+			"zone_name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+
+			"record": {
```
Comment (mbfrahry) [inline azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go:57]: Do you think any other attributes will be place under record? If not, we could probably shorten this to just be a TypeSet with string elements
```
Type:     schema.TypeSet,
					Required: true,
					Elem: &schema.Schema{
						Type: schema.TypeString,
					},
```

File: azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+	}
+
+	d.Set("name", name)
+	d.Set("resource_group_name", resGroup)
+	d.Set("zone_name", zoneName)
+	d.Set("ttl", resp.TTL)
+	d.Set("fqdn", resp.Fqdn)
+
+	if err := d.Set("record", flattenAzureRmPrivateDnsTxtRecords(resp.TxtRecords)); err != nil {
+		return err
```
Comment (mbfrahry) [inline azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go:166]: We should expand this out to:
```suggestion
		return fmt.Errorf("setting `record`: %s", err)
```
Just to give us a better picture of what thing we had trouble setting

File: azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+	return results
+}
+
+func expandAzureRmPrivateDnsTxtRecords(d *schema.ResourceData) *[]privatedns.TxtRecord {
+	recordStrings := d.Get("record").(*schema.Set).List()
+	records := make([]privatedns.TxtRecord, len(recordStrings))
+
+	segmentLen := 254
+	for i, v := range recordStrings {
+		record := v.(map[string]interface{})
```
Comment (mbfrahry) [inline azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go:219]: I'm always hesitant about making this type of assertion as I've seen cases where this is nil and we panic. Let's add a continue if this is nil and then we can keep the rest of the logic the same.

File: azurerm/internal/services/privatedns/tests/resource_arm_private_dns_txt_record_test.go [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+		CheckDestroy: testCheckAzureRMPrivateDnsTxtRecordDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAzureRMPrivateDnsTxtRecord_basic(data),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMPrivateDnsTxtRecordExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "record.#", "2"),
+				),
+			},
+			{
```
Comment (mbfrahry) [inline azurerm/internal/services/privatedns/tests/resource_arm_private_dns_txt_record_test.go:72]: Let's add import steps between all these checks and one more step at the end that takes us back to the basic configuration just to confirm it all works

File: azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go (hunk 1 above)
Comment (manicminer) [reply azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go:57]: Thanks, I wasn't sure about this actually as I wanted to try and keep it similar to the [azurerm_dns_txt_record](https://github.com/terraform-providers/terraform-provider-azurerm/blob/c916aa6db4142add1760a9b39bc668eb8b1dfcde/azurerm/internal/services/dns/resource_arm_dns_txt_record.go#L54-L66) resource for consistency. Do you think it's worth changing both resources (and deprecating the `value` key in the existing resource)?

File: azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go (hunk 3 above)
Comment (manicminer) [reply azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go:219]: Do you mean the record values may come back as nil? e.g. mitigate like:

```go
	segmentLen := 254
	for i, v := range recordStrings {
		if v == nil {
			continue
		}

		record := v.(map[string]interface{})
```

File: azurerm/internal/services/privatedns/tests/resource_arm_private_dns_txt_record_test.go (hunk 4 above)
Comment (manicminer) [reply azurerm/internal/services/privatedns/tests/resource_arm_private_dns_txt_record_test.go:72]: Great, thanks! Hoping I've interpreted your suggestion correctly, have pushed some changes.

File: azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go (hunk 3 above)
Comment (mbfrahry) [reply azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go:219]: Exactly. Looks great!

File: azurerm/internal/services/privatedns/tests/resource_arm_private_dns_txt_record_test.go [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+		Providers:    acceptance.SupportedProviders,
+		CheckDestroy: testCheckAzureRMPrivateDnsTxtRecordDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAzureRMPrivateDnsTxtRecord_withTags(data),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMPrivateDnsTxtRecordExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "tags.%", "2"),
+				),
+			},
```
Comment (mbfrahry) [inline azurerm/internal/services/privatedns/tests/resource_arm_private_dns_txt_record_test.go:105]: We want the import step here too

File: azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go (hunk 1 above)
Comment (mbfrahry) [reply azurerm/internal/services/privatedns/resource_arm_private_dns_txt_record.go:57]: Ahhh, yea. I get that. I think that's just leftover from trying to match the terraform schema to the api struct but it's not a best practice. I don't think we can match it at this time because we couldn't keep the same name. We could just note it as a todo for 3.0

**Comment by tombuildsstuff** [comment]
Tests pass:

Screenshot 2020-04-08 at 14 28 04
